lizmap/modules/lizmap/lib/Request/WFSRequest.php

```diff
@@ -346,6 +348,115 @@ protected function process_getfeature()
      */
     protected function getfeatureQgis()
     {
+        // In the WFS OGC standard FEATUREID and BBOX parameters cannot be mutually set
+        // but in Lizmap, the user can do a selection, based on featureid, and can request
+        // a download, a WFS GetFeature request, based on this selection with a restriction
+        // to map extent, sofeatureid and bbox parameter can be set mutually and featureid
+        // parameter needs to be transform in an expression filter.
+        // The transformation is only available if the QGIS layer has been set.
+        if ($this->param('featureid')
+            && $this->param('bbox')
+            && $this->qgisLayer) {
+            $provider = $this->qgisLayer->getProvider();
+
+            $typename = $this->requestedTypename();
+            $featureid = $this->param('featureid', '');
+
+            // The featureid is based on multi fields key
+            $hasDoubleAtSign = !empty(strstr($featureid, '@@'));
+
+            // Building the expression filter for single key
+            // or for multi fields key for postgres layer
+            $exp_filter = '';
+            if (($hasDoubleAtSign && $provider == 'postgres')
+                || !$hasDoubleAtSign) {
+                $fids = explode(',', $featureid);
+
+                // get primary keys values
+                $pks = array();
+                foreach ($fids as $fid) {
+                    $exp = explode('.', $fid);
+                    if (count($exp) == 2 and $exp[0] == $typename) {
```

Comment:
We should control which data is given here, because it comes from a parameter, and could be used for SQL injections. QGIS Server probably does the job, but we could perhaps use a small regex to check if the values is simple enough ?

Reply:
Wich regExp do you propose ?

Reply:
The method `validateFilter` seems to do the Job. Could you perhaps use it in this context ?

Reply:
`validateFilter is for postgres layer and update the expression to SQL, so I have introduced a `validateExpressionFilter` which is strict because of blockedWords. Key values cannot contain `;` or `--`.